### PR TITLE
refactor: refactor Info, replace PathBuf with Box<path>, delay .to_path_buf only when necessary

### PR DESCRIPTION
**Pull request by Boshen**
This PR greatly reduces PathBuf allocation by delaying calls to `to_path_buf`, using `Cow`, and using `Box<Path>`.

### CODE REVIEW

File: src/info.rs [hunk 1]
```diff
@@ -1,30 +1,42 @@
...
-    pub fn from(path: PathBuf, request: Request) -> Self {
-        Self { path, request }
+    pub fn new<P: AsRef<Path>>(path: P, request: Request) -> Self {
+        Self {
+            path: path.as_ref().into(),
+            request,
+        }
+    }
+
+    pub fn from_path<P: AsRef<Path>>(path: P) -> Self {
```
Comment (bvanjoi) [inline src/info.rs:22]: why not use `From` trait

File: src/info.rs [hunk 2]
```diff
@@ -33,6 +45,31 @@ impl Info {
...
+
+    #[must_use]
+    pub fn request(&self) -> &Request {
+        &self.request
+    }
+
+    #[must_use]
+    pub fn to_resolved_path(&self) -> Cow<'_, Path> {
+        if self.request.target().is_empty() || self.request.target() == "." {
+            Cow::Borrowed(&self.path)
```
Comment (bvanjoi) [inline src/info.rs:61]: great work

File: src/plugin/exports_field.rs [hunk 3]
```diff
@@ -22,17 +22,17 @@ impl<'a> ExportsFieldPlugin<'a> {
 impl<'a> Plugin for ExportsFieldPlugin<'a> {
     fn apply(&self, resolver: &Resolver, info: Info, context: &mut Context) -> State {
         // info.path should end with `node_modules`.
-        let target = &info.request.target();
+        let target = &info.request().target();
```
Comment (bvanjoi) [inline src/plugin/exports_field.rs:25]: remove ref symbol maybe better
```
let target = info.request().target();
```

Comment (Boshen) [reply src/plugin/exports_field.rs]: Great catch.

File: src/info.rs (hunk 1 above)
Comment (Boshen) [reply src/info.rs]: Great suggestion!

**Comment by Boshen** [comment]
Even for perf wins!

**Review by bvanjoi — APPROVED** [review]

**Pull request merged**